### PR TITLE
Speed up secondary invariants using ideas by Simon King

### PR DESCRIPTION
**Pull request by joschmitt**
Improves (hopefully) the computation of secondary invariants using "Fast computation of secondary invariants" by Simon King.

### CODE REVIEW

File: docs/oscar_references.bib [hunk 1]
```diff
@@ -812,6 +812,13 @@ @Article{Kem99
   url           = {https://doi.org/10.1006/jsco.1998.0247}
 }
 
+@Misc{Kin07,
+  author        = {King, Simon},
+  title         = {Fast computation of secondary invariants},
+  howpublished  = {arXiv:math/0701270},
+  year          = 2007
```
Comment (fingolfin) [inline docs/oscar_references.bib]: Please add the DOI, it is `10.48550/arXiv.math/0701270`

(We probably should review all our arXiv refs and make them uniform in this regard,

File: docs/oscar_references.bib [hunk 2]
```diff
@@ -1019,6 +1026,18 @@ @Book{Stu93
...
+@Article{Sym11,
+  author        = {Symonds, Peter},
+  title         = {On the {C}astelnuovo-{M}umford regularity of rings of
+                  polynomial invariants},
+  journal       = {Ann. of Math. (2)},
+  fjournal      = {Annals of Mathematics. Second Series},
+  volume        = {174},
+  year          = {2011},
+  number        = {1},
+  pages         = {499--517}
```
Comment (fingolfin) [inline docs/oscar_references.bib]: Please add the DOI `10.4007/annals.2011.174.1.14`

(Bib items with DOI automatically get a hyperlink to the paper in the Oscar manual)

File: experimental/InvariantTheory/primary_invariants.jl [hunk 3]
```diff
@@ -28,6 +21,27 @@ function test_primary_degrees_via_hilbert_series(R::InvRing, degrees::Vector{Int
...
+  g = denominator(mol)
+  for d in degrees
+    # multiply f by 1 - t^d
+    f -= shift_left(f, d)
+  end
+  if !is_zero(mod(f, g))
+    return false, parent(f)()
+  end
+  h = div(f, g)
+  return true, h
```
Comment (fingolfin) [inline experimental/InvariantTheory/primary_invariants.jl]: Isn't this the same as this?
```suggestion
  return divides(f, g)
```

File: experimental/InvariantTheory/secondary_invariants.jl [hunk 4]
```diff
@@ -217,10 +217,22 @@ function secondary_invariants_modular(RG::InvRing)
   s_invars_cache = SecondaryInvarsCache{elem_type(Rgraded)}()
   add_invariant!(s_invars_cache, one(Rgraded), false, Int[])
 
+  # The maximal degree in which we have to look for secondary invariants by Sym11.
```
Comment (fingolfin) [inline experimental/InvariantTheory/secondary_invariants.jl]: I was confused for a moment why a symmetric group of degree 11 comes up here ;-).

Perhaps like this:
```suggestion
  # The maximal degree in which we have to look for secondary invariants by [Sym11].
```
or
```suggestion
  # The maximal degree in which we have to look for secondary invariants by \cite{Sym11}.
```

**Review by fingolfin — COMMENTED** [review]
Nice!

You say "hopefully" improves performance -- have you don any benchmark tests at all?

**Comment by joschmitt** [comment]
> You say "hopefully" improves performance -- have you don any benchmark tests at all?

I'm fairly certain that there is no regression, but I'm not too sure that we gained much...
Anyways, there is this:
```
# master
julia> M1 = diagonal_matrix([ matrix(QQ, 3, 3, [ 0, 1, 0, 1, 0, 0, 0, 0, 1 ]) for i = 1:4 ]);
julia> M2 = diagonal_matrix([ matrix(QQ, 3, 3, [ 0, 1, 0, 0, 0, 1, 1, 0, 0 ]) for i = 1:4 ]);
julia> RG = invariant_ring(M1, M2);
julia> primary_invariants(RG);
julia> @btime Oscar.secondary_invariants_nonmodular(RG);
  2.909 s (4708433 allocations: 335.48 MiB)

# This branch
julia> @btime Oscar.secondary_invariants_nonmodular(RG);
  37.618 ms (265222 allocations: 21.92 MiB)
```

**Comment by fingolfin** [comment]
OK, that *does* look like a nice gain ;-). Besides, with the code in Julia, there's always the possibility to tweak it.

Thank you!

The PR is marked as draft, should I still try to review it, or do you plan further work before?

**Comment by joschmitt** [comment]
This is good to go now.

**Review by fingolfin — APPROVED** [review]

**Pull request merged**